### ISSUE TITLE
Compilation error on linux/PowerPC architecture

### ISSUE THREAD

**Issue by vzaliva** [coq/coq#10028]
I am trying to compile Coq-8.9.0 on Debian/PowerPC (big-endian architecture). I got the following error:

```
#=== ERROR while compiling coq.8.9.0 ==========================================#
# context     2.0.4 | linux/ppc32 | ocaml-system.4.05.0 | http://coq.inria.fr/opam/released#2019-04-27 07:10
# path        ~/.opam/default/.opam-switch/build/coq.8.9.0
# command     /usr/bin/make -j1
# exit-code   2
# env-file    ~/.opam/log/coq-2371-a00279.env
# output-file ~/.opam/log/coq-2371-a00279.out
### output ###
# [...]
# COQC      theories/Lists/ListTactics.v
# COQC      plugins/setoid_ring/InitialRing.v
# COQC      plugins/setoid_ring/Ring_tac.v
# File "./plugins/setoid_ring/Ring_tac.v", line 18, characters 0-21:
# Error:
# while loading
# /home/lord/.opam/default/.opam-switch/build/coq.8.9.0/plugins/quote/quote_plugin.cmxs:
# error loading shared library: /home/lord/.opam/default/.opam-switch/build/coq.8.9.0/plugins/quote/quote_plugin.cmxs: R_PPC_REL24 relocation at 0x0e6f0920 for symbol `caml_call_gc' out of range
# 
# make[1]: *** [plugins/setoid_ring/Ring_tac.vo] Error 1
# make[1]: Leaving directory `/home/lord/.opam/default/.opam-switch/build/coq.8.9.0'
# make: *** [submake] Error 2

```

Log files attached.

[coq-2371-a00279.out.gz](https://github.com/coq/coq/files/3129255/coq-2371-a00279.out.gz)
[coq-2371-a00279.env.gz](https://github.com/coq/coq/files/3129256/coq-2371-a00279.env.gz)

**Comment by ppedrot** [comment]
If anything, this looks more like a problem from the OCaml linking implementation.

**Comment by ejgallego** [comment]
Uh indeed, what does `ocamlc -config` output?

Can you try with 8.10, OCaml 4.07.1 and the Dune build?

**Comment by vzaliva** [comment]
```
$ ocamlc -config
version: 4.05.0
standard_library_default: /home/lord/ocamlbrew/ocaml-4.05.0/lib/ocaml
standard_library: /home/lord/ocamlbrew/ocaml-4.05.0/lib/ocaml
standard_runtime: /home/lord/ocamlbrew/ocaml-4.05.0/bin/ocamlrun
ccomp_type: cc
bytecomp_c_compiler: gcc -std=gnu99 -O2 -fno-strict-aliasing -fwrapv -fno-builtin-memcmp -D_FILE_OFFSET_BITS=64 -D_REENTRANT -fPIC
bytecomp_c_libraries: -lm  -ldl -lcurses -lpthread                  
native_c_compiler: gcc -std=gnu99 -O2 -fno-strict-aliasing -fwrapv -fno-builtin-memcmp -D_FILE_OFFSET_BITS=64 -D_REENTRANT
native_c_libraries: -lm  -ldl
native_pack_linker: ld -r -m elf32ppclinux  -o 
ranlib: ranlib
cc_profile: -pg
architecture: power
model: ppc
int_size: 31
word_size: 32
system: elf
asm: as -mppc
asm_cfi_supported: true
with_frame_pointers: false
ext_exe: 
ext_obj: .o
ext_asm: .s
ext_lib: .a
ext_dll: .so
os_type: Unix
default_executable_name: a.out
systhread_supported: true
host: powerpc-unknown-linux-gnu
target: powerpc-unknown-linux-gnu
profiling: true
flambda: false
spacetime: false
safe_string: false
exec_magic_number: Caml1999X011
cmi_magic_number: Caml1999I021
cmo_magic_number: Caml1999O011
cma_magic_number: Caml1999A012
cmx_magic_number: Caml1999Y015
cmxa_magic_number: Caml1999Z014
ast_impl_magic_number: Caml1999M020
ast_intf_magic_number: Caml1999N018
cmxs_magic_number: Caml2007D002
cmt_magic_number: Caml2012T009
```

I can try other versions of OCaml but it will take some time. It is QEMU VM with just 2Gb RAM (max allowed) and it is very slow. Rebuilding OCaml takes hours.

Frankly, right now this bug is not critical to me. I just noticed and reported this problem in case you want to know.

**Comment by ejgallego** [comment]
That looks like an OCaml bug, seeing the architecture you are using I am not surprised.

I dunno, if the machine doesn't allow debugging then maybe no point indeed on keeping this bug open. OCaml indeed is not supported in a few archs as of today, for example intel osx 32bit support is quite limited for Coq.

**Comment by vzaliva** [comment]
Same error with 4.07.1:

```
#=== ERROR while compiling coq.8.9.0 ==========================================#
# context     2.0.4 | linux/ppc32 | ocaml-base-compiler.4.07.1 | https://opam.ocaml.org#4b9884e4
# path        ~/.opam/4.07.1/.opam-switch/build/coq.8.9.0
# command     /usr/bin/make -j1
# exit-code   2
# env-file    ~/.opam/log/coq-9432-74f651.env
# output-file ~/.opam/log/coq-9432-74f651.out
### output ###
# [...]
# COQC      theories/Lists/ListTactics.v
# COQC      plugins/setoid_ring/InitialRing.v
# COQC      plugins/setoid_ring/Ring_tac.v
# File "./plugins/setoid_ring/Ring_tac.v", line 18, characters 0-21:
# Error:
# while loading
# /home/lord/.opam/4.07.1/.opam-switch/build/coq.8.9.0/plugins/quote/quote_plugin.cmxs:
# error loading shared library: /home/lord/.opam/4.07.1/.opam-switch/build/coq.8.9.0/plugins/quote/quote_plugin.cmxs: R_PPC_REL24 relocation at 0x0e6fe968 for symbol `caml_call_gc' out of range
# 
# make[1]: *** [plugins/setoid_ring/Ring_tac.vo] Error 1
# make[1]: Leaving directory `/home/lord/.opam/4.07.1/.opam-switch/build/coq.8.9.0'
# make: *** [submake] Error 2
```

However, according to this https://stackoverflow.com/questions/12862830/r-ppc-rel24-relocation-out-of-range the problem may not in OCaml but in compilation flags we use like missing `-fPIC`....

**Comment by ejgallego** [comment]
What does happens in `master` + `make -f Makefile.dune world`

**Comment by vzaliva** [comment]
I left it to compile overnight,  but after compiling ~10K files it failed:

```
$ make -f Makefile.dune world
dune build  @vodeps
dune exec ./tools/coq_dune.exe _build/default/.vfiles.d
dune build  @install
File "ide/dune", line 44, characters 12-22:
44 |  (libraries coqide_gui))
                 ^^^^^^^^^^
Error: Library "coqide_gui" in _build/default/ide is hidden (optional with unavailable dependencies).
Hint: try: dune external-lib-deps --missing @install
make: *** [world] Error 1  
```

It looks like dune build requires GTK libraries even when GUI is not built. 

```
$ dune external-lib-deps --missing @install
Error: The following libraries are missing in the default context:
- lablgtk3-sourceview3 (optional)
```

There is no `lablgtk3-sourceview3` under Debian Wheezy:

```
$ aptitude search lablgtk
p   liblablgtk-extras-ocaml-dev            - collection of modules for OCaml/LablGtk2 applicat
v   liblablgtk-extras-ocaml-dev-m3ke9      -                                                  
p   liblablgtk-extras-ocaml-doc            - collection of modules for OCaml/LablGtk2 applicat
p   liblablgtk2-gl-ocaml                   - runtime libraries for OCaml bindings to GtkGL    
v   liblablgtk2-gl-ocaml-0rem5             -                                                  
p   liblablgtk2-gl-ocaml-dev               - OCaml bindings to GtkGL                          
v   liblablgtk2-gl-ocaml-dev-0rem5         -                                                  
p   liblablgtk2-gnome-ocaml                - runtime libraries for OCaml bindings to GNOME    
p   liblablgtk2-gnome-ocaml-dev            - OCaml bindings to GNOME                          
v   liblablgtk2-gnome-ocaml-dev-f6nb5      -                                                  
v   liblablgtk2-gnome-ocaml-f6nb5          -                                                  
p   liblablgtk2-ocaml                      - runtime libraries for OCaml bindings for GTK+ ver
p   liblablgtk2-ocaml-dev                  - OCaml bindings to Gtk+ version 2                 
v   liblablgtk2-ocaml-dev-v69e2            -                                                  
p   liblablgtk2-ocaml-doc                  - OCaml bindings to Gtk+ version 2 (documentation) 
v   liblablgtk2-ocaml-v69e2                -                                                  
p   liblablgtkmathview-ocaml               - OCaml bindings for libgtkmathview (runtime)      
p   liblablgtkmathview-ocaml-dev           - OCaml bindings for libgtkmathview, a GTK widget t
v   liblablgtkmathview-ocaml-dev-uchx5     -                                                  
v   liblablgtkmathview-ocaml-uchx5         -                                                  
p   liblablgtksourceview2-ocaml            - OCaml bindings for libgtksourceview2 (runtime)   
p   liblablgtksourceview2-ocaml-dev        - OCaml bindings for libgtksourceview2, a source ed
v   liblablgtksourceview2-ocaml-dev-hzj30  -                                                  
v   liblablgtksourceview2-ocaml-hzj30      -      
```

**Comment by ejgallego** [comment]
Oh, so it seems it was working, you can compile without the GUI doing `make -f Makefile.dune coq`

**Comment by ejgallego** [comment]
But it seems to me that the compilation, other than for CoqIDE, succeed.

**Comment by vzaliva** [comment]
but did it link? Where I can check for an executable or shared lib which failed to build earlier?

**Comment by ejgallego** [comment]
That command should have compiled the whole stdlib so I guess it did link. You can check the build in `_build/install/`

**Comment by vzaliva** [comment]
I see `coqtop` and other binaries under `_build`, but no quote plugin, which compilation was caused the original error:

```
$ find _build/install/ -name \*quote\*
$ ls _build/install/default/lib/coq/plugins/
btauto      firstorder    ltac              omega        ssr              tauto
cc          fourier       micromega         r_syntax     ssreflect        tutorial
derive      funind        nsatz             rtauto       ssrmatching
extraction  int63_syntax  numeral_notation  setoid_ring  string_notation
$
```

**Comment by SkySkimmer** [comment]
We removed the quote plugin for 8.10.

**Comment by vzaliva** [comment]
Ok, I guess that solved the problem :)
I do not know if you want to keep this issue open or close it since it is fixed in 8.10. Up to you.
Thanks for sorting this out.

**Comment by ejgallego** [comment]
You are welcome!
